File: Java/src/main/java/QuestionBank/Solution201_300/Solution290.java
package QuestionBank.Solution201_300;

import java.util.HashMap;
import java.util.HashSet;

/**
 * @author windows
 */
public class Solution290 {
    public boolean wordPattern(String pattern, String s) {
        String[] strings = s.split(" ");
        if (pattern.length() != strings.length){
            return false;
        }
        HashMap<Character, String> map = new HashMap<>();
        HashSet<String> set = new HashSet<>();
        for (int i = 0; i < pattern.length(); i++) {
            if (map.containsKey(pattern.charAt(i))) {
                String str = map.get(pattern.charAt(i));
                if (!str.equals(strings[i])){
                    return false;
                }
            }else {
                if (set.contains(strings[i])){
                    return false;
                }
                map.put(pattern.charAt(i), strings[i]);
                set.add(strings[i]);
            }
        }
        return true;
    }



    public static void main(String[] args) {
        Solution290 solution290 = new Solution290();
    }
}

File: C++/src/Solution201_300/Solution290.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution290.h"
#include <unordered_map>

bool Solution290::wordPattern(string pattern, string s) {
    unordered_map<char, string> map1;
    unordered_map<string , char> map2;
    int i = 0;
    for (char c : pattern) {
        if (i >= s.length()) return false;
        int j = i;
        while (j < s.length() && s[j] != ' ') {
            j++;
        }
        string str = s.substr(i, j);
        i = j + 1;
        if (map1.count(c) && map1[c] != str) return false;
        if (map2.count(str) && map2[str] != c) return false;
        map1[c] = str;
        map2[str] = c;
    }
    return true;
}
